Support generating bootstrapping keys at the server side.

Guy-Moshkowich opened the pull request:
*Issue #22  

*Description of changes:*
This PR enable generating bootstrapping keys at the server side. Current API, requires the secret key which is not available in the server side. The new implementation require the generation of bootstrapping switching key pair and their marshaling.

1) Add support for un/marshal SwitchingKey 
2) Generate bootstrapping keys without secret key based on Lattigo v4.1.0 implementation https://github.com/tuneinsight/lattigo/blob/adf762375670e412bab261cd7ffff9ca03777ad5/ckks/bootstrapping/bootstrapper.go#L78
3) Enable generating conjugate rotation. Lattigo original API already supports this feature but the C++ wrapper used to call it with fixed "false" value. This is useful for generating bootstrapping keys manually.
4) Deprecated a function that accept unused parameters and fix bootstrapping example to use the fixed function

All the C++ wrapper examples are running successfully with this PR changes.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Guy-Moshkowich committed:
gen btpKey by implementing bootstrapping.GenEvaluationKeys in the wrapper. This is a first step to extract the secret key from gen hierarchical bts key

Guy-Moshkowich committed:
1) remove sk from gen hierarchical bts key

Guy-Moshkowich committed:
1) backward compatibility of genRotationKeysForRotations and genBootstrappingKey
2) add DEPRECATED description to legacy  genBootstrappingKey as it ignore two parameters.
3) update bootstrapping example to use new genBootstrappingKey api.

Guy-Moshkowich committed:
add marshal and unmarshal for SwitchingKey

Guy-Moshkowich committed:
improve deprecate description

crockeea commented on src/gowrapper/ckks/keygen.go:
```diff
@@ -336,7 +340,7 @@ func lattigo_setRotKeysForEvaluationKey(evalKeyHandle Handle5, rotKeysHandle Han
 
 //export lattigo_genBootstrappingKey
-func lattigo_genBootstrappingKey(keygenHandle Handle5, paramHandle Handle5, btpParamsHandle Handle5, skHandle Handle5, relinKeyHandle Handle5, rotKeyHandle Handle5) Handle5 {
+func lattigo_genBootstrappingKey(keygenHandle Handle5, paramHandle Handle5, btpParamsHandle Handle5, skHandle Handle5) Handle5 {
```
Nice catch.

crockeea commented on src/latticpp/ckks/keygen.cpp:
```diff
@@ -139,8 +143,22 @@ namespace latticpp {
 
-    BootstrappingKey genBootstrappingKey(const KeyGenerator &keygen, const Parameters &params, const BootstrappingParameters &bootParams, const SecretKey &sk, const RelinearizationKey &relinKey, const RotationKeys &rotKeys) {
-        return BootstrappingKey(lattigo_genBootstrappingKey(keygen.getRawHandle(), params.getRawHandle(), bootParams.getRawHandle(), sk.getRawHandle(), relinKey.getRawHandle(), rotKeys.getRawHandle()));
+    // DEPRACATED
+    BootstrappingKey genBootstrappingKey(const KeyGenerator &keygen, const Parameters &params, const BootstrappingParameters &bootParams, const SecretKey &sk, const RelinearizationKey & /*relinKey*/, const RotationKeys &/*rotKeys*/) {
+        return genBootstrappingKey(keygen, params, bootParams, sk);
+    }
```
I'm not terribly concerned with backwards compatibility; I suggest we just kill this function.
```suggestion
```

crockeea commented on src/latticpp/ckks/keygen.h:
```diff
@@ -63,6 +68,8 @@ namespace latticpp {
 
     RotationKeys genRotationKeysForRotations(const KeyGenerator &keygen, const SecretKey &sk, std::vector<int> shifts);
 
+    RotationKeys genRotationKeysForRotations(const KeyGenerator &keygen, const SecretKey &sk, std::vector<int> shifts, int includeConjugate);
```
Nit: We could just keep this API and set the default argument for `includeConjugate` to `false`.

crockeea commented on src/latticpp/ckks/keygen.h:
```diff
@@ -75,8 +82,15 @@ namespace latticpp {
     void setRotKeysForEvaluationKey(const EvaluationKey &evalKey,
                                     const RotationKeys &rotKeys);
 
+    // DEPRACATED: use genBootstrappingKey with the same list of parameters excluding RelinearizationKey, RotationKeys
```
```suggestion
```

crockeea reviewed (changes requested):
Thank you so much for this contribution!

Guy-Moshkowich committed:
fix code review comments:
1) remove of depractated genBootstrappingKey
2) remove non-relevant comment
3) set default value of paramaeter in genRotationKeysForRotations and remove redundant method overload.

Guy-Moshkowich commented:
Thank you for taking the time to review and for the good comments. I've fixed all of them.
